From 4d9990e94bef5555ec8e8350fdba186b4c3e42e2 Mon Sep 17 00:00:00 2001
From: Adam Spiers <aspiers@suse.com>
Date: Wed, 5 Nov 2014 00:42:30 +0100
Subject: [PATCH] fix chef-client to work when cluster is down (bnc#900964)

We need chef-client to work on a cluster node even when openais is down,
so that Chef can restart openais.  The start_handler was carelessly
checking for maintenance mode in a way which assumed openais was up.

https://bugzilla.suse.com/show_bug.cgi?id=900964
---
 .../libraries/maintenance_mode_helpers.rb            | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/chef/cookbooks/crowbar-pacemaker/libraries/maintenance_mode_helpers.rb b/chef/cookbooks/crowbar-pacemaker/libraries/maintenance_mode_helpers.rb
index d471b76..0484582 100644
--- a/chef/cookbooks/crowbar-pacemaker/libraries/maintenance_mode_helpers.rb
+++ b/chef/cookbooks/crowbar-pacemaker/libraries/maintenance_mode_helpers.rb
@@ -19,6 +19,26 @@ module CrowbarPacemaker
   # Chef::Provider::PacemakerService LWRP.
   module MaintenanceModeHelpers
     def maintenance_mode?
+      # For once, we want 2>&1 to come before >/dev/null, not after!
+      cibadmin = %x(cibadmin -Ql 2>&1 >/dev/null)
+      case cibadmin
+      when /Connection refused/
+        # Cluster is not up, so let things proceed so that Chef can
+        # start it if appropriate.
+        return false
+      when /command not found/
+        # pacemaker has been deinstalled?
+        return false
+      end
+
+      if ! $?.success?
+        Chef::Log.warn("cibadmin -Ql failed when checking Pacemaker maintenance mode!")
+        Chef::Log.warn(cibadmin)
+        Chef::Log.warn("Something wrong, so treating as if in maintenance " +
+          "mode; will need manual intervention.")
+        return true
+      end
+
       !! (%x(crm node show #{node.hostname}) =~ /maintenance:\s*on/)
     end
 
-- 
1.8.4.5

